[WIP] - Referenceable Object Swift PropertyWrapper - @FirestoreObjectReference

aashishpatil-g opened the pull request:
### Discussion
This PR proposes a Swift property wrapper (@FirestoreObjectReference) to store object (non-scalar) properties (vars) by reference in Firestore instead with the document blob. It automates storing the object in a specified location and saves the reference id in the parent document instead of the actual object. This is useful for scenarios where a common object must be used / referenced across multiple parent objects.  

When reading/loading the parent document (object) from Firestore, the propertyWrapper provides a way to load the referenced object via the projected value of the wrapper. The loadReference() call on the projected value is async.

For objects to participate in the propertyWrapper, they must implement the ReferenceableObject Protocol. 

Here is a sample data model and how the propertyWrapper is used. 

  We have three structs representing a simplified UserProfile model -
  `UserProfile`
  `Employer` - an employer who the `UserProfile` works for
  `WorkLocation` - representing a generic location. 
 
  Since multiple Users can work for an Employer, it makes sense to have only one instance of an Employer that is referred to
  by multiple Users. Similarly multiple users can be in a WorkLocation. Additionally, an Employer can also be located
  in a WorkLocation (e.g. headquarters of an Employer). We can mark WorkLocation and Employer as a ReferenceableObjects.
 
 
  ```
  struct UserProfile: ReferenceableObject {
      var username: String
 
      @FirestoreObjectReference
      var employer: Employer?
 
      @FirestoreObjectReference
      var workLocation: WorkLocation?
 
  }
  
  struct Employer: ReferenceableObject {
      var name: String
 
      @FirestoreObjectReference
      var headquarters: WorkLocation?
  }
 
  struct WorkLocation: ReferenceableObject {
      var locationName: String
      var moreInfo: String //
 
  }
 
  var userProfile = ...
 
  // use projected value to load referenced employer
  try await userProfile.$employer?.loadObject()
 
  // use projected value to load referenced workLocation
  try await prof.$workLocation?.loadObject()
 
```

 
### TODO
The following items are still to be done.
  * Provide an example app. Have to clean up what I have
  * Unit Tests

peterfriese commented on Firestore/Swift/Source/ReferenceableObject/FirestoreObjRefWrapper.swift:
```diff
@@ -0,0 +1,150 @@
+
+/// Property wrapper @FirestoreObjectReference,
+/// Indicates that the specified property value should be stored by reference instead of by value,  inline
+/// with the parent object.
```
I am not sure what "inline with the parent object" refers to.

aashishpatil-g replied:
I can't find a good way to describe this. Checked both these videos but not finding a good naming.

Firestore data structure
https://www.youtube.com/watch?v=haMOUb3KVSo

Firestore Data Modeling I/O 2019 
https://www.youtube.com/watch?v=lW7DWV2jST0

peterfriese commented on Firestore/Swift/Source/ReferenceableObject/FirestoreObjRefWrapper.swift:
```diff
@@ -0,0 +1,150 @@
+/// When loading a parent object, any references are not loaded by default and can be loaded on demand
+/// using the projected value of the wrapper.
+///
+/// structs that can be stored as a reference must implement the `ReferenceableObject` protocol
```
Uppercase?

aashishpatil-g replied:
Fixed.

peterfriese commented on Firestore/Swift/Source/ReferenceableObject/FirestoreObjRefWrapper.swift:
```diff
@@ -0,0 +1,150 @@
+///
+/// structs that can be stored as a reference must implement the `ReferenceableObject` protocol
+///
+/// variables that are annotated with the propertyWrapper should be marked as `Optional` since they can be nil when not loaded or set
```
Uppercase

aashishpatil-g replied:
Fixed

peterfriese commented on Firestore/Swift/Source/ReferenceableObject/FirestoreObjRefWrapper.swift:
````diff
@@ -0,0 +1,150 @@
+///
+/// ```
+/// struct UserProfile: ReferenceableObject {
+///     var username: String
````
Indentation 2 spaces?

aashishpatil-g replied:
Fixed

peterfriese commented on Firestore/Swift/Source/ReferenceableObject/FirestoreObjRefWrapper.swift:
```diff
@@ -0,0 +1,150 @@
+///
+@available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
+@propertyWrapper
+public struct FirestoreObjRef<T> where T: ReferenceableObject {
```
I know it's quite a mouthful, but I think I might prefer FirestoreObjectReference over the shortened version. Code completion is your friend.

aashishpatil-g replied:
Makes sense. Made the change. Tx!

peterfriese commented on Firestore/Swift/Source/ReferenceableObject/ReferenceableObject.swift:
```diff
@@ -0,0 +1,138 @@
+/// in the Employer collection and a reference to Employer is stored within the UserProfile encoded document.
+///
+@available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
+public struct ObjReference<T: ReferenceableObject>: Codable {
```
ObjectReference

aashishpatil-g replied:
Changed

peterfriese commented on Firestore/Swift/Source/ReferenceableObject/FirestoreObjRefWrapper.swift:
```diff
@@ -0,0 +1,150 @@
+
+  private mutating func updateInitialValue(initialValue: T?) {
+    if var initialValue {
+      var oid = initialValue.id
```
What does oid mean?

aashishpatil-g replied:
short for objectId. I expanded it out.

peterfriese commented on Firestore/Swift/Source/ReferenceableObject/FirestoreObjRefWrapper.swift:
```diff
@@ -0,0 +1,150 @@
+        initialValue.id = oid
+      }
+      objReference = ObjReference(
+        documentId: oid!,
```
Can we write this differently, to avoid force unwrapping? Maybe using a guard statement to replace the if on line 90?

aashishpatil-g replied:
Good catch! Made the change. 

I need to check and set if not set. 
Can't use guard since it requires a return. Instead used the optional coalescing operator.

peterfriese commented on Firestore/Swift/Source/ReferenceableObject/FirestoreObjRefWrapper.swift:
```diff
@@ -0,0 +1,150 @@
+
+  public var wrappedValue: T? {
+    get {
+      return objReference?.referencedObj
```
Again, I'd prefer objectReference and referencedObject.

peterfriese commented on Firestore/Swift/Source/ReferenceableObject/ReferenceableObject.swift:
```diff
@@ -0,0 +1,138 @@
+@available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
+public protocol ReferenceableObject: Codable, Identifiable, Hashable, Equatable {
+  // The Firestore collection where objects of this type are stored.
+  // If no value is specified, it defaults to Type name of the object
```
"defaults to Type name of" -> "defaults to the type name of"

peterfriese commented on Firestore/Swift/Source/ReferenceableObject/ReferenceableObject.swift:
```diff
@@ -0,0 +1,138 @@
+
+  // Helper function that creates a path for the object
+  static func objectPath(objectId: String) -> String {
+    let coll = Self.parentCollection()
```
coll -> collection

aashishpatil-g replied:
Changed

peterfriese commented on Firestore/Swift/Source/ReferenceableObject/ReferenceableObject.swift:
```diff
@@ -0,0 +1,138 @@
+
+// MARK: Contained Object Reference
+
+/// Struct used to store a reference to a ReferenceableObject. When a container Struct is
```
"container Struct" -> "container struct"

aashishpatil-g replied:
changed

peterfriese commented on Firestore/Swift/Source/ReferenceableObject/ReferenceableObject.swift:
```diff
@@ -0,0 +1,138 @@
+
+/// Struct used to store a reference to a ReferenceableObject. When a container Struct is
+/// encoded, any FirestoreObjRef property wrappers are encoded as ObjReference objects and the referenced
+/// objects are stored (if needed) in their intended location instead of inline.
```
"inline" - I haven't seen this terminology being used in the context of Firestore. Check with the Firestore team what their preferred way of describing this is.

aashishpatil-g replied:
I couldn't figure out a better way of saying this. Let me see some docs too.

peterfriese commented on Firestore/Swift/Source/ReferenceableObject/ReferenceableObject.swift:
```diff
@@ -0,0 +1,138 @@
+  public var collection: String
+  public var placeholder: Codable?
+
+  ///
```
Missing comment?

aashishpatil-g replied:
Added comment

peterfriese commented on Firestore/Swift/Source/ReferenceableObject/ReferenceableObject.swift:
```diff
@@ -0,0 +1,138 @@
+  public var referencedObj: T?
+
+  enum CodingKeys: String, CodingKey {
+    case documentId = "docId"
```
Is `docId` a name that is used by Firestore by default?

aashishpatil-g replied:
Makes sense - I changed it to use objectId so its consistent with this PR.

peterfriese commented on Firestore/Swift/Source/ReferenceableObject/ReferenceableObjectManager.swift:
```diff
@@ -0,0 +1,257 @@
+      let encoder = Firestore.Encoder()
+      let json = try encoder.encode(object)
+
+      if let docId = object.id {
```
documentId

aashishpatil-g replied:
Changed

peterfriese commented on Firestore/Swift/Source/ReferenceableObject/ReferenceableObjectManager.swift:
```diff
@@ -0,0 +1,257 @@
+  public func save<T: ReferenceableObject>(object: T) async throws {
+    do {
+      let encoder = Firestore.Encoder()
+      let json = try encoder.encode(object)
```
Why don't we use Firestore's Codable support to encode this?

aashishpatil-g replied:
I changed it in the else condition. If the if block, the code has to generate checksum so the code anyway needs to generate the json. To avoid re-generating within that method I directly used the FirestoreEncoder here.

peterfriese commented on Firestore/Swift/Source/ReferenceableObject/ReferenceableObjectManager.swift:
```diff
@@ -0,0 +1,257 @@
+        try await db.collection(T.parentCollection()).document().setData(json)
+      }
+
+      print("Save complete")
```
Might want to use the logger.

aashishpatil-g replied:
Done. I also moved the refactored the Logger class so it could be reused across FirestoreSwift if need be.

peterfriese commented on Firestore/Swift/Source/ReferenceableObject/ReferenceableObjectManager.swift:
```diff
@@ -0,0 +1,257 @@
+    do {
+      // first check cache
+      if let cacheEntry = await objectCache.get(for: T.objectPath(objectId: objId)) {
+        return cacheEntry.object as! T
```
Is there a way how we can cast the type without coercing it?

aashishpatil-g replied:
I haven't been able to figure this one out. Its odd. 'object' is a ReferenceableObject and so is 'T'. Not sure why compiler is not letting me just return object without a cast.

aashishpatil-g replied:
Ok optional casting removes the warning and its not a forced cast anymore. In reality if cacheEntry exists, object will not be nil since its not an optional but optional casting makes compiler happy.

peterfriese commented on Firestore/Swift/Source/ReferenceableObject/ReferenceableObjectManager.swift:
```diff
@@ -0,0 +1,257 @@
+      }
+
+      // get from db
+      let docRef = db.collection(T.parentCollection()).document(objId)
```
documentReference / objectId

aashishpatil-g replied:
Changed

peterfriese commented on Firestore/Swift/Source/ReferenceableObject/ReferenceableObjectManager.swift:
```diff
@@ -0,0 +1,257 @@
+    return foundObjects
+  }
+
+  public func fetchObjects<T: ReferenceableObject>(predicates: [QueryPredicate]) async throws
```
Our other APIs use `get` instead of `fetch` and `Document` instead of `Object`,  (see https://firebase.google.com/docs/firestore/query-data/get-data#get_a_document)

aashishpatil-g replied:
Changed the function name to use get. Kept object to stay consistent with the PRs nomenclature.

peterfriese commented on Firestore/Swift/Source/ReferenceableObject/ReferenceableObjectManager.swift:
```diff
@@ -0,0 +1,257 @@
+
+    var foundObjects = [T]()
+    let snapshot = try await query.getDocuments()
+    for document in snapshot.documents {
```
Maybe use a compactMap instead (like we do in most of our samples for mapping collections).

aashishpatil-g replied:
This doesn't seem to work. I think the issue is that its within an async method.

peterfriese reviewed (changes requested):
I generally like the idea (object references are a feature of Firestore, but we haven't had client side support for them so far), however, I would love to hear the Firestore team's opinion on whether we should expose this to the client SDKs.

**Sample app**: I started a sample at in https://github.com/firebase/firebase-ios-sdk/tree/master/Example/FirestoreSample, which contains a menu structure that should make it easy to add a sample for this property wrapper.

aashishpatil-g commented:
@peterfriese 

Thanks for the detailed review. I've addressed the comments except
- Not sure whats the right way to describe the "inline storage" vs "by reference"
- Will do the sample app in a separate commit.

paulb777 commented:
Please close, merge, or comment. We plan to close stale PRs on November 28, 2023.

aashishpatil-g commented:
Going to close this for now as I haven't made progress on sample app. Will reopen later.
